controllers/argocd/notifications.go

```diff
@@ -0,0 +1,428 @@
+		return err
+	}
+
+	return nil
+}
+
+// The code to create/delete notifications resources is written within the reconciliation logic itself. However, these functions must be called
+// in the right order depending on whether resources are getting created or deleted. During creation we must create the role and sa first.
+// RoleBinding and deployment are dependent on these resouces. During deletion the order is reversed.
+// Deployment and RoleBinding must be deleted before the role and sa. deleteNotificationsResources will only be called during
+// delete events, so we don't need to worry about duplicate, recurring reconciliation calls
+func (r *ReconcileArgoCD) deleteNotificationsResources(cr *argoprojv1a1.ArgoCD) error {
+
+	sa := &corev1.ServiceAccount{}
+	role := &rbacv1.Role{}
+
+	if err := argoutil.FetchObject(r.Client, cr.Namespace, fmt.Sprintf("%s-%s", cr.Name, common.ArgoCDNotificationsControllerComponent), sa); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	if err := argoutil.FetchObject(r.Client, cr.Namespace, fmt.Sprintf("%s-%s", cr.Name, common.ArgoCDNotificationsControllerComponent), role); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	log.Info("reconciling notifications deployment")
+	if err := r.reconcileNotificationsDeployment(cr, sa); err != nil {
+		return err
+	}
+
+	log.Info("reconciling notifications role binding")
+	if err := r.reconcileNotificationsRoleBinding(cr, role, sa); err != nil {
+		return err
+	}
+
+	log.Info("reconciling notifications serviceaccount")
+	_, err := r.reconcileNotificationsServiceAccount(cr)
+	if err != nil {
+		return err
+	}
+
+	log.Info("reconciling notifications role")
+	_, err = r.reconcileNotificationsRole(cr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileArgoCD) reconcileNotificationsServiceAccount(cr *argoprojv1a1.ArgoCD) (*corev1.ServiceAccount, error) {
+
+	sa := newServiceAccountWithName(common.ArgoCDNotificationsControllerComponent, cr)
+
+	saExists := true
+	if err := argoutil.FetchObject(r.Client, cr.Namespace, sa.Name, sa); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get the serviceAccount associated with %s : %s", sa.Name, err)
+		}
+		saExists = false
+	}
+
+	if saExists {
+		// SA exists but shouldn't, so it should be deleted
+		if !cr.Spec.Notifications.Enabled {
+			log.Info(fmt.Sprintf("Deleting serviceaccount %s as notifications is disabled", sa.Name))
+			return nil, r.Client.Delete(context.TODO(), sa)
+		}
+		return sa, nil
+	}
+
+	// SA doesn't exist and shouldn't, nothing to do here
+	if !cr.Spec.Notifications.Enabled {
+		return nil, nil
+	}
+
+	// SA doesn't exist but should, so it should be created
+	if err := controllerutil.SetControllerReference(cr, sa, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	log.Info(fmt.Sprintf("Creating serviceaccount %s", sa.Name))
+	err := r.Client.Create(context.TODO(), sa)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Can we move this inside `FetchObject` call before updating value of `saExists = false` (line 104) and update the workflow accordingly?

Reply:
I'm not sure I follow
what would the benefit of doing that be?

Reply:
It is more like, if fetching sa object errors out, we know that we do not have an existing sa. So we just create it there itself and return rather than performing other set of checks.

Also, we have followed similar approach everywhere in the repo.

Reply:
If you're referring to the workflow that is followed in reconciling the default resources (like server, repo-server etc) maybe in that case a check is not required because we need those resources to exist on the cluster under all circumstances as long as the owner Argo CD instance is around
but notifications resources are optional, so I think we should be more deliberate about when they should be created and make sure they are cleaned up when not required